src/test/java/com/google/starfish/services/FavoriteNoteServiceTest.java

```diff
@@ -0,0 +1,300 @@
+
+import com.google.starfish.services.FavoriteNoteService.Recency;
+import com.google.starfish.models.Note;
+import com.ninja_squad.dbsetup.DbSetup;
+import com.ninja_squad.dbsetup.destination.DataSourceDestination;
+import com.ninja_squad.dbsetup.operation.Operation;
+
+@RunWith(JUnit4.class)
+public class FavoriteNoteServiceTest {
+  private Constants constants = new Constants();
+  private String NOTES = constants.NOTES;
+  private String LABELS = constants.LABELS;
+  private String FAVORITE_NOTES = constants.FAVORITE_NOTES;
+  private String USERS = constants.USERS;
+  private DataSource pool = Constants.pool;
+  private FavoriteNoteService favoriteNoteService = new FavoriteNoteService();
+
+  private final Date today = favoriteNoteService.getDateBasedOnRecency(Recency.TODAY);
+  private final Date thisWeek = favoriteNoteService.getDateBasedOnRecency(Recency.THIS_WEEK);
+  private final Date thisMonth = favoriteNoteService.getDateBasedOnRecency(Recency.THIS_MONTH);
+  private final Date allTime = favoriteNoteService.getDateBasedOnRecency(Recency.ALL_TIME);
+
+  private final String newSchool = "cornell university";
+  private final String newCourse = "cs1300";
+  private final String firstNewUserId = Integer.toString(Integer.parseInt(constants.REFERENCE_USER_ID) + 1);
+  private final String secondNewUserId = Integer.toString(Integer.parseInt(constants.REFERENCE_USER_ID) + 2);
+
+  private static final boolean runTests = Constants.TEST_DB_NAME.equals("starfish_test");
+
+  /** Clear DB and insert reference data */
+  @Before
+  public void prepare() throws Exception {
+    if(!runTests) throw new Exception("Wrong Test Database Name");
+    Operation operation = getBeforeTestOperation();
+    DbSetup dbSetup = new DbSetup(new DataSourceDestination(pool), operation);
+    dbSetup.launch();
+  }
+
+  /** Test to see if we can get all notes favorited by a single user */
+  @Test
+  public void testGettingFavoriteNotesByUserId() throws SQLException {
+    Note[] favoriteNotes = favoriteNoteService.getFavoriteNotesByUserId(pool, firstNewUserId);
+    assertEquals(favoriteNotes.length, 7);
+  }
+
+  /** Test to see if we can get the number of times a particular note has been favorited */
+  @Test
+  public void testGettingNumFavoritesByNoteId() throws SQLException {
+    long numFavoritesOfNoteOne = favoriteNoteService.getNumFavoritesByNoteId(pool, 1);
+    long numFavoritesOfNoteTwo = favoriteNoteService.getNumFavoritesByNoteId(pool, 2);
+    assertTrue(numFavoritesOfNoteOne == 1);
+    assertTrue(numFavoritesOfNoteTwo == 2);
+  }
+
+  /** Test to see if we can correctly retrieve the most trending notes today */
+  @Test
+  public void testGettingTrendingNotesToday() throws SQLException, Exception {
+    Object[][] trendingNotesToday = favoriteNoteService.getTrendingNotesBySchoolOrCourse(pool, Recency.TODAY, null, null);
+    // We should be getting all 7 notes irrespective of whether they are trending or not
```

Comment:
(Here and below) I'm a bit confused by this statement that you should be getting *all* notes irrespective of whether they're trending or not-- perhaps the name of the method then is misleading because you say you're getting trending notes, specifically?  Also, aren't you necessarily limiting the number of notes you're getting by having a date limit, so why would you be getting all?

Reply:
I believe we introduced this functionality to the `getTrendingNotesBySchoolOrCourse` method to work as we expected with the search servlet. Basically, irrespective of the timespan somebody is seeking "trending" notes, we want the backend to return all the note data sorted by the number of favorites in that timespan, and any note that isn't "trending" would have a number of favorites in that timespan of 0, but would still be returned because it would still technically fit the user's search query. Like, this method isn't being used in the context of a user searching specifically for trending notes in a given timespan, but a user just searching for notes in general and this method allows us to internally rank those notes. The notes that aren't "trending" just move to the bottom of the array and then it's really easy for the frontend to filter those notes out by removing notes with a 0 count on num favorites in the timespan if that's the functionality desired there.